File: shoppingOffers/shoppingOffers.java
class Solution {
    public int shoppingOffers(List<Integer> price, List<List<Integer>> special, List<Integer> needs) {
        return dfs(price, special, needs, 0);
    }
    
    public int dfs(List<Integer> price, List<List<Integer>> special, List<Integer> needs, int start) {
        int res =  purchaseWithOriginalPrice(price, needs);
        for (int i = start; i < special.size(); i++) {
            List<Integer> offer = special.get(i);
            List<Integer> temp = new ArrayList<>();
            for (int j = 0; j < needs.size(); j++) {
                if (needs.get(j) < offer.get(j)) { // check if able to buy one offer
                    temp = null;
                    break;
                }
                temp.add(needs.get(j) - offer.get(j)); // buy j
            }
            if (temp != null) { // can buy one offer
                res = Math.min(res, offer.get(offer.size() - 1) + dfs(price, special, temp, i));
            }
        }
        return res;
    }
    
    public int purchaseWithOriginalPrice(List<Integer> price, List<Integer> needs) {
        int total = 0;
        for (int i = 0; i < needs.size(); i++) {
            total += price.get(i) * needs.get(i);
        }
        return total;
    }
}

File: 638.shoppingOffers/shoppingOffers.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int shoppingOffers(vector<int>& price, vector<vector<int>>& special, vector<int>& needs) {
        return dfs(price, special, needs, 0);
    }
    int dfs(vector<int>& price, vector<vector<int> >& special, vector<int> needs, int index) {
        if (index == special.size()) {
            return purchaseWithOrginalPrice(price, needs);
        }
        vector<int> nextNeeds = needs;
        int i;
        for (i = 0; i < price.size(); i++) {   // try to buy special[index]
            int remain = needs[i] - special[index][i];
            if (remain < 0) // can not buy
                break;
            else
                nextNeeds[i] = remain;
        }
        if (i == price.size()) // can buy one special[index]
            return min (special[index][i] + dfs(price, special, nextNeeds, index),
                       dfs(price, special, needs, index + 1));
        else
            return dfs(price, special, needs, index + 1);
    }
    int purchaseWithOrginalPrice(vector<int>& price, vector<int>& needs) {
        int sum = 0;
        for (int i = 0; i < needs.size(); i++) {
            sum += needs[i] * price[i];
        }
        return sum;
    }
};

int main() {
  Solution s;
  vector<int> price = {2,3,4};
  vector<vector<int> > special = {{1,1,0,4}, {2,2,1,9}};
  vector<int> needs = {1,2,1};
  cout << s.shoppingOffers(price, special, needs) << endl;
  return 0;
}
